code/game/machinery/spaceheater.dm

```diff
@@ -60,128 +129,142 @@
 					C.loc = src
 					C.add_fingerprint(usr)
 
-					user.visible_message("[user] inserts a power cell into [src].", "<span class='notice'>You insert the power cell into [src].</span>")
+					user.visible_message("\The [user] inserts a power cell into \the [src].", "<span class='notice'>You insert the power cell into \the [src].</span>")
+					SSnano.update_uis(src)
 		else
 			user << "<span class='warning'>The hatch must be open to insert a power cell!</span>"
 			return
 	else if(istype(I, /obj/item/weapon/screwdriver))
-		open = !open
-		user.visible_message("[user] [open ? "opens" : "closes"] the hatch on \the [src].", "<span class='notice'>You [open ? "open" : "close"] the hatch on \the [src].</span>")
+		panel_open = !panel_open
+		user.visible_message("\The [user] [panel_open ? "opens" : "closes"] the hatch on \the [src].", "<span class='notice'>You [panel_open ? "open" : "close"] the hatch on \the [src].</span>")
 		update_icon()
-		if(!open && user.machine == src)
-			user << browse(null, "window=spaceheater")
-			user.unset_machine()
+		if(panel_open)
+			interact(user)
+	else if(exchange_parts(user, I) || default_deconstruction_crowbar(I))
+		return
 	else
 		..()
 
-
 /obj/machinery/space_heater/attack_hand(mob/user)
-	src.add_fingerprint(user)
-	if(open)
-
-		var/dat
-		dat = "Power cell: "
-		if(cell)
-			dat += "<A href='byond://?src=\ref[src];op=cellremove'>Installed</A><BR>"
-		else
-			dat += "<A href='byond://?src=\ref[src];op=cellinstall'>Removed</A><BR>"
-
-		dat += "Power Level: [cell ? round(cell.percent(),1) : 0]%<BR><BR>"
-
-		dat += "Set Temperature: "
-
-		dat += "<A href='?src=\ref[src];op=temp;val=-5'>-</A>"
-
-		dat += " [set_temperature]&deg;C "
-		dat += "<A href='?src=\ref[src];op=temp;val=5'>+</A><BR>"
-
-		user.set_machine(src)
-		user << browse("<HEAD><TITLE>Space Heater Control Panel</TITLE></HEAD><TT>[dat]</TT>", "window=spaceheater")
-		onclose(user, "spaceheater")
-
-
+	if(!user)
+		return
+	interact(user)
 
+/obj/machinery/space_heater/attack_paw(mob/user)
+	if(!user)
+		return

```

Comment:
Are those two `if(!user)` check necessary? If some code actually manages to call attack_hand with a null user it'd be a bug and we shouldn't try to hide it.